### PR TITLE
[GCS]Do not save task spec in GCS actor table(Part1)

### PR DESCRIPTION
**Pull request by ffbin**
## Why are these changes needed?
In the 200 node cluster test, we found that after GCS restart, it takes a long time for GCS client to fetch actor information from GCS server, because the task spec is large and the serialization time is long. And GCS will mistakenly think that the node heartbeat timeout.
Currently, task spec is only required when GCS creates or restarts an actor. Publishing an actor and getting an actor/actors do not need to return a task spec, so we store task spec in actor task spec table separately.
This feature is split into three PR:
Currently, it is the first PR, and the second PR supports synchronous acquisition of actor task spec data. The third PR deals with the recovery process after GCS restart.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
Is it ready for review? Can you resolve the merge conflict and remove the tag in that case?

**Comment by ffbin** [comment]
> Is it ready for review? Can you resolve the merge conflict and remove the tag in that case?

Thanks, i will resolve the merge conflict. I find that we can't remove task_spec directly because the new dashboard depends on it.

**Comment by rkooo567** [comment]
What are values that the dashboard depends on? We can probably extract them when creating an Actor class in gcs_actor_manager.cc. I am sure only a few fields are used for the dashboard, and it is probably waste to keep it just for the dashboard.

**Comment by ffbin** [comment]
> What are values that the dashboard depends on? We can probably extract them when creating an Actor class in gcs_actor_manager.cc. I am sure only a few fields are used for the dashboard, and it is probably waste to keep it just for the dashboard.

image
Does removing task_spec affect `get_actor_creation_tasks` function?

**Comment by rkooo567** [comment]
@ffbin 
```python3
def actor_classname_from_task_spec(task_spec):
    return task_spec.get("functionDescriptor", {})\
                .get("pythonFunctionDescriptor", {})\
                .get("className", "Unknown actor class")
```
We only use these 3 fields, so keeping the whole task spec would be wasteful?

**Comment by rkooo567** [comment]
But cc @mfitton just in case

**Comment by mfitton** [comment]
@rkooo567 @ffbin Sang is right that we only need those fields from the task spec, but we pull that data from the `GetNodeStatsReply` protobuf (defined in `node_manager.proto`), so some kind of task spec data (even if it has unnecessary fields removed) will need to remain in that reply. This data is currently supplied in the `NodeManager` function `HandleGetNodeStats` taken from the `local_queues_` attribute.

If this data is removed from the reply, we will no longer be able to show users information about pending and infeasible actors within the logical view.

**Comment by ffbin** [comment]
> @rkooo567 @ffbin Sang is right that we only need those fields from the task spec, but we pull that data from the `GetNodeStatsReply` protobuf (defined in `node_manager.proto`), so some kind of task spec data (even if it has unnecessary fields removed) will need to remain in that reply. This data is currently supplied in the `NodeManager` function `HandleGetNodeStats` taken from the `local_queues_` attribute.
> 
> If this data is removed from the reply, we will no longer be able to show users information about pending and infeasible actors within the logical view.
> 
> Correct me if I'm wrong because I only glanced through the PR, but it doesn't seem like this affects the data stored in the local task queue per node at all, so it shouldn't actually affect the dashboard.

Thank you for your detailed explains :)

File: src/ray/protobuf/gcs.proto
Comment (rkooo567) [inline src/ray/protobuf/gcs.proto:150]: Can we take out stats to a separate message? Maybe call it ActorStats?

File: src/ray/protobuf/gcs.proto
Comment (rkooo567) [inline src/ray/protobuf/gcs.proto:45]: Btw, internal config was 29 before lol,

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:76]:
```suggestion
  RAY_CHECK(task_spec_) << "This method shouldn't be called after the ClearCreationTaskSpecification method is invoked"; 
```

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:441]: If an actor is dead here, there should be no problem right? (I cannot really think of issues here, but I'd like to know your thought!)

**Review by rkooo567 — COMMENTED** [review]
LGTM in general. Just a couple comments. 

Btw, does it mean after this PR is merged actor restart or gcs restart will still work? Or should we wait until 3 PRs are merged?

**Comment by ffbin** [comment]
> LGTM in general. Just a couple comments.
> 
> Btw, does it mean after this PR is merged actor restart or gcs restart will still work? Or should we wait until 3 PRs are merged?

After this PR is merged, gcs restart will not work until 3 PRs are merged, thanks.

**Comment by rkooo567** [comment]
@ffbin Can you add skip for gcs server restart tests?

**Comment by ffbin** [comment]
> @ffbin Can you add skip for gcs server restart tests?

Oh, i will add skip for gcs server restart tests, thanks.

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:441]: There should be a problem. If the actor is dead, we may override the actor state. I think I need to add a judgment on whether the actor is dead before put ActorTable, thanks.

File: src/ray/protobuf/gcs.proto
Comment (ffbin) [reply src/ray/protobuf/gcs.proto:45]: Changing its value should have no effect? Otherwise, TABLE_PREFIX_MAX will be smaller than INTERNAL_CONFIG.

File: src/ray/protobuf/gcs.proto
Comment (ffbin) [reply src/ray/protobuf/gcs.proto:150]: Hi @rkooo567 , Which variables do you want to put in ActorStats?

**Comment by ffbin** [comment]
Hi @rkooo567 , Can you help review this PR, thanks :)

File: python/ray/tests/test_gcs_fault_tolerance.py
Comment (rkooo567) [inline python/ray/tests/test_gcs_fault_tolerance.py:50]: Is it true?

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:329]: Can you explain why in the comment?

File: src/ray/protobuf/gcs.proto
Comment (rkooo567) [reply src/ray/protobuf/gcs.proto:43]: I meant it is a good catch. I am surprised how it worked before.

File: src/ray/protobuf/gcs.proto
Comment (rkooo567) [reply src/ray/protobuf/gcs.proto:148]: I thought we pull actor stats data from the gcs, but it is probably not true. Please ignore this comment.

File: python/ray/tests/test_gcs_fault_tolerance.py
Comment (rkooo567) [reply python/ray/tests/test_gcs_fault_tolerance.py:50]: Oops nvm. I re-read the description.

File: src/ray/protobuf/gcs.proto
Comment (ffbin) [reply src/ray/protobuf/gcs.proto:148]: Ok, thanks

**Review by rkooo567 — APPROVED** [review]

**Comment by ffbin** [comment]
Hi @mfitton , I found that this PR will affect `new_dashboard:test_logical_view_head`, can you help have a look?  thanks https://travis-ci.com/github/ray-project/ray/jobs/454201286

**Comment by kfstorm** [comment]
@ffbin Any progress on this PR?

**Comment by ffbin** [comment]
> @ffbin Any progress on this PR?

This PR will affect `new_dashboard:test_logical_view_head`, the problem has not yet been solved.

**Comment by rkooo567** [comment]
cc @mfitton

**Comment by rkooo567** [comment]
Hey @ffbin do you have a plan to merge this soon?

**Comment by ffbin** [comment]
> Hey @ffbin do you have a plan to merge this soon?

Yeah, this will bring great benefits to the performance. I will see how to solve the problem of dashboard.

**Comment by rkooo567** [comment]
cc @ffbin any status on this PR?

**Comment by ffbin** [comment]
@rkooo567 Sorry, I haven't invested in this PR recently. I will ask @clay4444  to help me finish it.

**Comment by rkooo567** [comment]
cc @ffbin no plan to work on this further? Should I close it?

**Comment by kfstorm** [comment]
@clay4444 Could you update the status?

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by WangTaoTheTonic** [comment]
I'll pick this up.

**Comment by kfstorm** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```

**Pull request closed without merging**
